Allow Plutus cost calculation to take offline data instead of querying the node

palas opened the pull request:
# Changelog

```yaml
- description: |
    Modified Plutus cost calculation command to allow the user to supply offline data instead of querying the node
  type:
  - feature
```

# Context

This PR addresses this issue: https://github.com/IntersectMBO/cardano-cli/issues/945
This PR depends on the following PR from `cardano-api`, I would recommend looking at both in conjunction: https://github.com/IntersectMBO/cardano-api/pull/771

# How to trust this PR

The test should provide some confidence, there is not a lot of change in the actual implementation, except that the parameters allow to provide the required info directly instead of querying the node.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

palas committed:
Add query to obtain `EraHistory`

Make `--out-file` optional

Add support for eras before conway

palas committed:
Redesign plutus cost calculation params to support offline

palas committed:
Add offline test for `calculate-plutus-script-cost` command

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Command.hs:
```diff
@@ -253,6 +255,13 @@ data QueryTreasuryValueCmdArgs era = QueryTreasuryValueCmdArgs
   deriving Show
 
+data QueryEraHistoryCmdArgs era = QueryEraHistoryCmdArgs
+  { eon :: !(ConwayEraOnwards era)
```
Why is this restricted to Conway?

palas replied:
It is because I am using `EraParamsWithGenesisWindow` parameter which has this comment [here](https://agrius.feralhosting.com/palas/docs/ouroboros-consensus-0.22.0.0-df65a647587553a959a08c5217cce9b83665426cae2920a78101c1f183df5c69/share/doc/html/src/Ouroboros.Consensus.HardFork.History.EraParams.html#EraParamsWithGenesisWindow):
```haskell
-- | Older versions are not aware of the Genesis window as part of 'EraParams',
-- so we need to stay backwards-compatible for now. This type can be removed
-- once mainnet is in Conway (as we can then always use the behavior of
-- 'EraParamsWithGenesisWindow').
```

So I thought it would be safer.

palas replied:
But maybe it is already not necessary to worry about it and just use `EraParamsWithGenesisWindow`

Jimbo4350 replied:
It doesn't need to be. This query can be in any shelley based era.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/80dbabb4e7ceebccbf89c5d408185b0d5bda3007

Jimbo4350 commented on cardano-cli/test/cardano-cli-golden/files/golden/help.cli:
```diff
@@ -7963,6 +7964,19 @@ Usage: cardano-cli conway query proposals [--cardano-mode [--epoch-slots SLOTS]]
+                                              [--volatile-tip | --immutable-tip]
+                                              --out-file FILEPATH
+
+  Obtains the era history data. Era history data can be used to calculate slot
```
This needs a little more fleshing out. Why is calculating slot times useful? Where will this be used?

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/fcf3d2a9d1657a4f01e1aad62b1d53335f21ff3e

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -243,11 +246,22 @@ data TransactionCalculateMinValueCmdArgs era = TransactionCalculateMinValueCmdAr
   , outputFile :: !(Maybe (File () Out))
   }
 
+data NodeContextInfo
```
These types are  missing haddocks explaining what this type is.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/c722a04785977e6437d69fa41544c7abb139bf43

Jimbo4350 replied:
Thoughts on `NodeContextInfoSource`?

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/0e4186f2727a1bb67feafcb3245f1ab05cc2c5cc

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1692,7 +1717,8 @@ runTransactionCalculatePlutusScriptCostCmd
     calculatePlutusScriptsCosts
-      :: CardanoEra era
+      :: forall era
+       . CardanoEra era
```
Something is wrong if `CardanoEra era` exists here. Plutus scripts become available in Alonzo.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/8ef0d86c26e8b88fb5c4393a254856d393edaf02

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -375,20 +376,77 @@ pTransactionCalculateMinReqUTxO era' =
+pSystemStart :: Parser SystemStart
+pSystemStart = systemStartUTC <|> systemStartPOSIX
+
+systemStartPOSIX :: Parser SystemStart
```
This can be read from the genesis file we should also give that option.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/685d8c83f831316dabe3261b1f13f96047ca9731

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -257,11 +258,16 @@ data NodeContextInfo
   , protocolParamsFile :: ProtocolParamsFile
   }
 
+data MustExtendSafeZone
```
Another type missing haddocks explaining in detail what this is.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/c722a04785977e6437d69fa41544c7abb139bf43

Jimbo4350 replied:
I would copy the definition here: https://github.com/IntersectMBO/cardano-cli/pull/1079/files?diff=unified&w=0#diff-53110ad7a92a5cf0005a13b3cdebb6d2103748bfdf04632d04a45380079a49f0R423

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/6c495875d6b314eb6a8ea8686ae192432f66592a

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -393,11 +393,27 @@ pNodeConnectionInfo =
+    ( mconcat
+        [ Opt.long "unsafe-extend-safe-zone"
+        , Opt.help
+            ( "Allow extending the validity of the era history past the safe zone "
```
This means nothing to an incoming developer. What is the safe zone? What is the validity of the safe zone? What is the era history?

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/c050f04ba792f8939cf1a1e2fc6bcd84637769f6

Jimbo4350 commented on cardano-cli/cardano-cli.cabal:
```diff
@@ -362,6 +362,8 @@ test-suite cardano-cli-test
     filepath,
     hedgehog,
     hedgehog-extras,
+    lens,
```
Why not `microlens`? I want to avoid pulling in `lens` or `lens-aeson`.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/8ab0791ada386b0e90c546bca1ed3801259c2024

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Option.hs:
```diff
@@ -340,6 +340,7 @@ pQueryCmds era envCli =
     , pQueryTreasuryValueCmd era envCli
     , pQueryProposalsCmd era envCli
     , pQueryStakePoolDefaultVote era envCli
+    , pQueryEraHistoryCmd era envCli
```
It would be nice to have these sorted in ascending order.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/c5888429bc40182d2543370d121907815ee57e3b

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Option.hs:
```diff
@@ -818,3 +819,26 @@ pQueryCommons w envCli =
+          <$> pQueryEraHistoryCmdArgs
+      )
+    $ Opt.progDesc
+      ( "Obtains the era history data. The era history contains information about when era transitions happened and can "
```
Use `mconcat []` instead of `<>`.

Jimbo4350 replied:
Nice description :+1: .

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/d0aa78f4efb9847ca5255525d8974edd81f319ed

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -245,11 +250,45 @@ data TransactionCalculateMinValueCmdArgs era = TransactionCalculateMinValueCmdAr
+  { systemStartSource :: SystemStartOrGenesisFile
+  , mustExtendSafeZone :: MustExtendSafeZone
+  , eraHistoryFile :: File EraHistory In
+  , utxoFile :: FilePath
```
We should also use `File` here.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/c304c9c4c526dc47d4c4338a80914e717739c0c2

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -245,11 +250,45 @@ data TransactionCalculateMinValueCmdArgs era = TransactionCalculateMinValueCmdAr
+  }
+
+-- | The system start time or a means to get it (the genesis file)
+data SystemStartOrGenesisFile
```
We should stick with the `*Source` naming convention.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/68b55aa87e3044c0db198e0296482b17830520cd

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -245,11 +250,45 @@ data TransactionCalculateMinValueCmdArgs era = TransactionCalculateMinValueCmdAr
+-- is run, or information required to obtain it (information to connect to the node).
+data NodeContextInfo
+  = NodeConnectionInfo !LocalNodeConnectInfo
+  | TransactionContextInfo !TransactionContext
```
`ProvidedTransactionContextInfo`? Naming is difficult here.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/2aca81a36e9c1f55a3ec48f083b471f8e5ce7e8b

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -391,17 +392,104 @@ pTransactionCalculatePlutusScriptCost :: EnvCli -> Parser (TransactionCmds era)
+    ( mconcat
+        [ Opt.long "unsafe-extend-safe-zone"
+        , Opt.help
+            ( "Allow extending the validity of the era history past the safe zone. The "
```
use `mconcat []`

Jimbo4350 replied:
> and we have confidence that the conversion from slot numbers to POSIX times using the era history will be correct.

We can improve this by saying we are confident that the slot _duration_ will not change.

Jimbo4350 replied:
>   "This safe zone is conservative, and if the user knows that a hard fork hasn't happened since the era history was obtained, it is safe for it to be used past the safe zone by using this flag."

This sentence is not clear. The era history has been obtained but it is safe to be used past the safe zone by using this flag? I don't understand.

palas replied:
`mconcat` applied here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/d0aa78f4efb9847ca5255525d8974edd81f319ed

palas replied:
Improved explanation here: https://github.com/IntersectMBO/cardano-cli/commit/fa214dcd2768cf903d2cf4491f27fe3247152225

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -391,17 +392,104 @@ pTransactionCalculatePlutusScriptCost :: EnvCli -> Parser (TransactionCmds era)
+  (SystemStartLiteral <$> (systemStartUTC <|> systemStartPOSIX))
+    <|> ( SystemStartFromGenesisFile . GenesisFile
+            <$> parseFilePath
+              "start-time-from-byron-genesis-file"
```
why not just `genesis-file`? We can indicate in the help that we are only extracting the start time.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/5f16fa2672b2eea1938fe47a104f487f79fd5f4b

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -391,17 +392,104 @@ pTransactionCalculatePlutusScriptCost :: EnvCli -> Parser (TransactionCmds era)
+pSystemStart :: Parser SystemStartOrGenesisFile
+pSystemStart =
+  (SystemStartLiteral <$> (systemStartUTC <|> systemStartPOSIX))
+    <|> ( SystemStartFromGenesisFile . GenesisFile
```
you can use `asum`. Makes aligning the code easier.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/29467111fdac3de2c13a4ba1597c1bc486e3d413

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -391,17 +392,104 @@ pTransactionCalculatePlutusScriptCost :: EnvCli -> Parser (TransactionCmds era)
+  File
+    <$> parseFilePath
+      "era-history-file"
+      ( "Filepath of the era history file as produced by the 'query era-history' command. "
```
`mconcat []`

Jimbo4350 replied:
Nice explanation :+1:

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/d0aa78f4efb9847ca5255525d8974edd81f319ed

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -391,17 +392,104 @@ pTransactionCalculatePlutusScriptCost :: EnvCli -> Parser (TransactionCmds era)
+pUtxoFile :: Parser FilePath
+pUtxoFile =
+  parseFilePath "utxo-file" $
+    "Filepath of the JSON-encoded file with info about the set of relevant "
```
"Filepath to a JSON-encoded UTxO file produced by the 'query utxo' command."

palas replied:
I think the bit about "relevant UTxOs" is important, because having the whole UTxO is a lot to ask. So I will write:
"Filepath to a JSON-encoded UTxO file as produced by the 'query utxo' command. Only UTxOs referenced by the transaction are needed, not the whole UTxO, but unnecessary info will be ignored."

palas replied:
Let me know if that works

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/de92320b5ae775e09ae14e0243d53be80b43324d

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1687,28 +1719,35 @@ runTransactionCalculatePlutusScriptCostCmd
-          (\aeo -> pure $ collectPlutusScriptHashes aeo tx txEraUtxo)
-          & hoistMaybe (TxCmdAlonzoEraOnwardsRequired era')
+    calculatePlutusScriptsCosts aeo systemStart eraHistory pparams txEraUtxo tx = do
+      let era' :: CardanoEra era = toCardanoEra aeo
```
Unnecessary `forall era`. If you remove the type signature from `era'` you can remove the `forall`.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/83d5b180a82dea30a2e18764ff704fab04f0dfbb

Jimbo4350 reviewed (commented):
Nice work. After this round of commits and tidying up the commit history this is good to go.

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Command.hs:
```diff
@@ -264,6 +266,13 @@ data QueryStakePoolDefaultVoteCmdArgs era = QueryStakePoolDefaultVoteCmdArgs
+data QueryEraHistoryCmdArgs era = QueryEraHistoryCmdArgs
+  { sbe :: !(ShelleyBasedEra era)
+  , commons :: !QueryCommons
+  , outFile :: !(File () Out)
```
Why it's not `Maybe`? Let's stick to the convention and allow stdout output.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/dc74b0da91db62fca7761286221abda64304ef4e

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -391,17 +392,104 @@ pTransactionCalculatePlutusScriptCost :: EnvCli -> Parser (TransactionCmds era)
+              ]
+        )
+
+systemStartUTC :: Parser SystemStart
```
Is there a benefit for having separate flags? Why not have a single flag and try to parse POSIX seconds and then fall back to predefined format? Seems more intuitive to me this way.

(This is similar to how [`-d` in `date`](https://www.gnu.org/software/coreutils/manual/html_node/Options-for-date.html) is implemented, so there's a precedent).

palas replied:
As of this particular case it is quite doable. But in general, I think the advantage is that we can give a more accurate error, instructions, and get rid of ambiguity.

If we want to later support POSIX in milliseconds, for example, by having one single option it would be a problem. Or if we want to support US and EU date formats at the same time, USA is MM-DD-YYYY, EU is DD-MM-YYYY, so 03-03-2003 is ambiguous.

But still, if the user writes:
`20230305123453`, did they miss the `-` and `:`, or did they write it in milliseconds, or is the number to big for a POSIX time? We can't really know

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1662,23 +1676,41 @@ runTransactionCalculatePlutusScriptCostCmd
+                eEraHistory <- queryEraHistory
+                eeUtxo <- queryUtxo txEra (QueryUTxOByTxIn relevantTxIns)
+                ePp <- queryExpr $ QueryInEra $ QueryInShelleyBasedEra sbe Api.QueryProtocolParameters
+                return $ do
```
what's the purpose of this nested do?

palas replied:
Combining the `Either`'s, the inner `do` is an `Either` monad, the outter one is `LocalStateQueryExpr`

palas replied:
I just realised there are a couple of unnecessary `do`s though. Removed them in: https://github.com/IntersectMBO/cardano-cli/commit/3db121679a0042c62965598ae2b25dbc370628ac

carbolymer replied:
Ah right. Error handling in `LocalStateQueryExpr` is an eyesore - not your fault though.

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1687,28 +1719,35 @@ runTransactionCalculatePlutusScriptCostCmd
-      txEraUtxo
-      tx
+    caseByronOrShelleyBasedEra
+      (left $ TxCmdAlonzoEraOnwardsRequired nodeEra)
```
This command should be limited to `AlonzoEraOnwards era` instead of reporting an error for previous eras.

palas replied:
The era is obtained from the transaction and the node. The error is when both the node and the transaction are in Byron. I can ask for a witness from the command group, but that is only going to add a new kind of error (mismatch with transaction and node), and I don't see this command appearing in anything Byron related judging by the changes to the golden files

carbolymer replied:
Fair point, I've misread the code.

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1738,6 +1777,44 @@ runTransactionCalculatePlutusScriptCostCmd
+        let systemStartUTCTime = Byron.gdStartTime byronGenesisData
+        return $ SystemStart systemStartUTCTime
+    utxosBytes <- handleIOExceptT (TxCmdUTxOFileError . FileIOError utxoFile) $ BS.readFile utxoFile
+    utxos <- firstExceptT TxCmdUTxOJSONError $ ExceptT (return $ Aeson.eitherDecodeStrict' utxosBytes)
```
```suggestion
    utxos <- liftEither . first TxCmdUTxOJSONError $ Aeson.eitherDecodeStrict' utxosBytes
```

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/c304c9c4c526dc47d4c4338a80914e717739c0c2

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1738,6 +1777,44 @@ runTransactionCalculatePlutusScriptCostCmd
+        (byronGenesisData, _) <- firstExceptT TxCmdGenesisDataError $ Byron.readGenesisData byronGenesisFile
+        let systemStartUTCTime = Byron.gdStartTime byronGenesisData
+        return $ SystemStart systemStartUTCTime
+    utxosBytes <- handleIOExceptT (TxCmdUTxOFileError . FileIOError utxoFile) $ BS.readFile utxoFile
```
If `utxoFile` were `File _` then:
```suggestion
    utxosBytes <- modifyError TxCmdUTxOFileError $ readByteStringFile utxoFile
```

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/c304c9c4c526dc47d4c4338a80914e717739c0c2

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1738,6 +1777,44 @@ runTransactionCalculatePlutusScriptCostCmd
+    EraHistory interpreter <-
+      onLeft (left . TxCmdTextEnvError) $
+        liftIO $
+          readFileTextEnvelope (proxyToAsType (error "Proxy type for EraHistory evaluated")) eraHistoryFile
```
Why not? Why this error has to be here?
```suggestion
          readFileTextEnvelope (proxyToAsType Proxy) eraHistoryFile
```

palas replied:
Oh, I didn't see that. Yes, pretty obvious in retrospective :/
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/4948776d92d01182be8fc615bec74d782281f25a
Thank you

carbolymer commented on cardano-cli/src/Cardano/CLI/Type/Error/TxCmdError.hs:
```diff
@@ -94,6 +97,9 @@ data TxCmdError
   | TxCmdUnregisteredStakeAddress !(Set StakeCredential)
   | forall era. TxCmdAlonzoEraOnwardsRequired !(CardanoEra era)
+  | TxCmdUTxOFileError !(FileError JsonDecodeError)
+  | TxCmdUTxOJSONError String
```
camel casing is wrong here, it should be:
```suggestion
  | TxCmdUtxoFileError !(FileError JsonDecodeError)
  | TxCmdUtxoJsonError String
```

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/51c4f09d339fc9fdc4dedff90a9ff29a5bb6294b

carbolymer commented on cardano-cli/src/Cardano/CLI/Type/Error/TxCmdError.hs:
```diff
@@ -245,6 +251,12 @@ renderTxCmdError = \case
+  TxCmdUTxOFileError e ->
+    "Error while reading UTxO set from JSON file: " <> prettyError e
+  TxCmdUTxOJSONError e ->
+    "Error while reading UTxO set from JSON file: " <> pretty e
```
Why two separate errors if they mean the same? `TxCmdUTxOJSONError` constructor name should reflect it's purpose if you're stating it in the error message.

palas replied:
Mainly because the error info inside is different, but I've made the error message a bit more precise here: https://github.com/IntersectMBO/cardano-cli/commit/296022750a4e35f77ce07b8dfe2a671877e3fac8

newhoggy commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Option.hs:
```diff
@@ -215,50 +215,23 @@ pQueryCmds era envCli =
+    , pQueryDRepStakeDistributionCmd era envCli
+    , pQueryEraHistoryCmd era envCli
+    , pQueryFuturePParamsCmd era envCli
+    , pQueryGetGovStateCmd era envCli
```
Thank you for doing the change.

newhoggy commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Option.hs:
```diff
@@ -215,50 +215,23 @@ pQueryCmds era envCli =
-                  ]
+          commandWithMetavar "kes-period-info" $
+            Opt.info (pKesPeriodInfoCmd era envCli) $
+              Opt.progDesc "Get information about the current KES period and your node's operational certificate."
```
I discovered that if you use `.` instead of `$` except for the last one, `fourmolu` does something different:

```
    , Just
        . Opt.hsubparser
        . commandWithMetavar "kes-period-info"
        . Opt.info (pKesPeriodInfoCmd era envCli)
        $ Opt.progDesc "Get information about the current KES period and your node's operational certificate."
```

This seems to be to be tidier, and less indenting means there is more horizontal space for the description.

Thoughts?

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/641c18f893f76061e2940ebb0f067221c5220e21

newhoggy commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Option.hs:
```diff
@@ -818,3 +819,27 @@ pQueryCommons w envCli =
+          <$> pQueryEraHistoryCmdArgs
+      )
+    $ Opt.progDesc
+    $ mconcat
```
I wonder why `fourmolu` hasn't aggressively indented this one.  Is it because of the `info`?

newhoggy replied:
Ah yes, if I make the `info` call one line `fourmolu` does this:

```
  Just $
    Opt.hsubparser $
      commandWithMetavar "era-history" $
        Opt.info (QueryEraHistoryCmd <$> pQueryEraHistoryCmdArgs) $
          Opt.progDesc $
            mconcat
              [ "Obtains the era history data. The era history contains information about when era transitions happened and can "
              , "be used together with the start time to convert slot numbers to POSIX times offline (without connecting to the node). "
              , "Converting slot numbers to POSIX times is useful, for example, when calculating the cost of executing a Plutus "
              , "script. And being able to do it offline means that it can be calculated without access to a live node."
              ]
```

Then switching to `.` makes it vertical again:

```
  Just
    . Opt.hsubparser
    . commandWithMetavar "era-history"
    . Opt.info (QueryEraHistoryCmd <$> pQueryEraHistoryCmdArgs)
    . Opt.progDesc
    $ mconcat
      [ "Obtains the era history data. The era history contains information about when era transitions happened and can "
      , "be used together with the start time to convert slot numbers to POSIX times offline (without connecting to the node). "
      , "Converting slot numbers to POSIX times is useful, for example, when calculating the cost of executing a Plutus "
      , "script. And being able to do it offline means that it can be calculated without access to a live node."
      ]
```

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/641c18f893f76061e2940ebb0f067221c5220e21

newhoggy commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -19,6 +20,10 @@ module Cardano.CLI.EraBased.Transaction.Command
+  , NodeContextInfoSource (..)
+  , TransactionContext (..)
+  , MustExtendSafeZone (..)
+  , SystemStartOrGenesisFileSource (..)
```
Might be good to move the types together and sort them.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/82ce2698572ac80f106fca3255b673786a6a6008

newhoggy commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -245,11 +250,48 @@ data TransactionCalculateMinValueCmdArgs era = TransactionCalculateMinValueCmdAr
+  = NodeConnectionInfo !LocalNodeConnectInfo
+  | ProvidedTransactionContextInfo !TransactionContext
+
+-- | Transaction context, requried to evaluate the execution
```
`requried` typo.

palas replied:
Nice catch! Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/32939f4da07617dd263bed0737ec6b95b5e63e8e

newhoggy commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -245,11 +250,48 @@ data TransactionCalculateMinValueCmdArgs era = TransactionCalculateMinValueCmdAr
+  , protocolParamsFile :: ProtocolParamsFile
+  }
+
+-- | The system start time or a means to get it (the genesis file)
```
"The system start time or the genesis file from which to get it"

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/b3bc70dea9db4ec1eba733a79586a8f0b0ce55e3

newhoggy commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -391,17 +392,116 @@ pTransactionCalculatePlutusScriptCost :: EnvCli -> Parser (TransactionCmds era)
+    <$> ( Opt.strOption $
+            mconcat
+              [ Opt.long "start-time-utc"
+              , Opt.metavar "UTC-TIME"
```
Optimally, we should use underscore instead of hyphen for metavar.

carbolymer replied:
Yes, we're using screaming snake case for metavars.

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/914fa6b0d93514ff2f51e9e3c272abf3b57936ac

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1738,6 +1776,44 @@ runTransactionCalculatePlutusScriptCostCmd
+  -> SystemStartOrGenesisFileSource
+  -> MustExtendSafeZone
+  -> File EraHistory In
+  -> File () In
```
```suggestion
  -> File (UTxO era) In
```
if it doesn't feel like shoehorning

palas replied:
I left it out because it is a bit tricky since I didn't have the era in the parameters. But I may as well bite the bullet, I added a casting function for it so that I don't have to add a new error: https://github.com/IntersectMBO/cardano-cli/commit/6dad8d77521b9cc5195845707ae7237d813b4db7

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1687,28 +1719,34 @@ runTransactionCalculatePlutusScriptCostCmd
+      (left $ TxCmdAlonzoEraOnwardsRequired nodeEra)
+      ( caseShelleyToMaryOrAlonzoEraOnwards
+          (\_ -> left $ TxCmdAlonzoEraOnwardsRequired nodeEra)
+          ( \aeo ->
```
```suggestion
    aeo <- forEraMaybeEon nodeEra & hoistMaybe (TxCmdAlonzoEraOnwardsRequired nodeEra)
```

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/commit/a091928dd26950e9c90aa6e3d25e8610489da4f2

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Option.hs:
```diff
@@ -818,3 +819,22 @@ pQueryCommons w envCli =
+pQueryEraHistoryCmd w envCli =
+  Just
+    $ Opt.hsubparser
+    $ commandWithMetavar "era-history"
```
This isn't showing up in the top level `query` command and it should be:
```
cabal run exe:cardano-cli -- query                                                                                                          :( 1 25-04-16 - 9:43:30
Configuration is affected by the following files:
- cabal.project
- cabal.project.local
Resolving dependencies...
Created semaphore called cabal_semaphore_k with 16 slots.
Usage: cardano-cli query 
                           ( protocol-parameters
                           | tip
                           | stake-pools
                           | stake-distribution
                           | stake-address-info
                           | utxo
                           | ledger-state
                           | protocol-state
                           | stake-snapshot
                           | leadership-schedule
                           | kes-period-info
                           | pool-state
                           | tx-mempool
                           | slot-number
                           | ledger-peer-snapshot
                           )

  Node query commands. Will query the local node whose Unix domain socket is
  obtained from the CARDANO_NODE_SOCKET_PATH environment variable.

Available options:
  -h,--help                Show this help text

Available commands:
  protocol-parameters      Get the node's current protocol parameters
  tip                      Get the node's current tip (slot no, hash, block no)
  stake-pools              Get the node's current set of stake pool ids
  stake-distribution       Get the node's current aggregated stake distribution
  stake-address-info       Get the current delegations and reward accounts
                           filtered by stake address.
  utxo                     Get a portion of the current UTxO: by tx in, by
                           address or the whole.
  ledger-state             Dump the current ledger state of the node
                           (Ledger.NewEpochState -- advanced command)
  protocol-state           Dump the current protocol state of the node
                           (Ledger.ChainDepState -- advanced command)
  stake-snapshot           Obtain the three stake snapshots for a pool, plus the
                           total active stake (advanced command)
  pool-params              DEPRECATED. Use query pool-state instead. Dump the
                           pool parameters
                           (Ledger.NewEpochState.esLState._delegationState._pState._pParams
                           -- advanced command)
  leadership-schedule      Get the slots the node is expected to mint a block in
                           (advanced command)
  kes-period-info          Get information about the current KES period and your
                           node's operational certificate.
  pool-state               Dump the pool state
  tx-mempool               Local Mempool info
  slot-number              Query slot number for UTC timestamp
  ledger-peer-snapshot     Dump the current snapshot of big ledger peers. These
                           are the largest pools that cumulatively hold 90% of
                           total stake.
```

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/6b44663441e812693876eef6c3817a9d5f2fb132

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Query/Command.hs:
```diff
@@ -330,7 +330,7 @@ renderQueryCmds = \case
     "query stake-pool-default-vote"
   QueryEraHistoryCmd{} ->
-    "interpreter"
+    "query era-history"
```
:+1:

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/Compatible/Transaction/Option.hs:
```diff
@@ -62,7 +62,7 @@ pTxInOnly =
     (readerFromParsecParser parseTxIn)
     ( Opt.long "tx-in"
-        <> Opt.metavar "TX-IN"
+        <> Opt.metavar "TX_IN"
```
Can you create an ADR surrounding this change? I think this is the 3rd time we're changing this.

palas replied:
Created issue here: https://github.com/IntersectMBO/cardano-cli/issues/1148

Jimbo4350 commented on cardano-cli/test/cardano-cli-golden/files/golden/help/latest_genesis_create-staked.cli:
```diff
@@ -31,7 +31,7 @@ Available options:
   --gen-stake-delegs INT   The number of stake delegator credential sets to make
                            [default is 0].
-  --start-time UTC-TIME    The genesis start time in YYYY-MM-DDThh:mm:ssZ
+  --start-time UTC_TIME    The genesis start time in YYYY-MM-DDThh:mm:ssZ
```
Remove this change from this PR and open an ADR for it.

palas replied:
Removed it from this commit: https://github.com/IntersectMBO/cardano-cli/commit/513622cc826f0f7cce3e90ec8a57af83f797368e

palas replied:
Created issue here: https://github.com/IntersectMBO/cardano-cli/issues/1148

Jimbo4350 reviewed (changes requested):
- No need for the `metavar` changes to be in this PR. Please remove them to a separate PR and open an ADR.
- This needs to be addresses: https://github.com/IntersectMBO/cardano-cli/pull/1079#discussion_r2046975603
- Introduce online vs offline as pointed out in slack

 Everything else looks good. Happy to approve after.

palas committed:
Improvements to design and code style

- Split online and offline functionality into subcommands
- Use point free style to improve code layout
- Replace `evalEitherM` and `readJsonFile` with `readJsonFileOk`
- Remove parametrisation on era

Co-authored-by: Mateusz Galazyn
<228866+carbolymer@users.noreply.github.com>

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -433,13 +433,16 @@ pMustExtendEraHistorySafeZone =
+                <> "This safe zone is conservative. Even if we are past the safe zone, if "
+                <> "there hasn't been any era transition (hard fork) since we obtained it, we can "
+                <> "continue safely using the era history. "
+                <> "This flag essentially disables the safe zone check. This allows the user to "
```
:+1:

Jimbo4350 commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Run.hs:
```diff
@@ -1789,12 +1775,15 @@ runTransactionCalculatePlutusScriptCostCmd
           )
         $ encodePretty scriptCostOutput
 
+    castUtxoFileEra :: File (UTxO era1) In -> File (UTxO era2) In
```
We really shouldn't need to be casting eras and this an be a potential source of ambiguous errors if we cast from an older to a newer era.

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -400,21 +401,133 @@ pTransactionCalculateMinReqUTxO era' =
+  ( Opt.hsubparser $
+      commandWithMetavar "online" $
+        Opt.info (pTransactionCalculatePlutusScriptCostParams (pNodeConnectionInfo envCli)) $
+          Opt.progDesc
```
suggestion: use leading `.` instead of trailing `$` to vertically align the code with fourmolu

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/ca0ab1fffb1b78353424921a47c30bdee50f9b56

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Option.hs:
```diff
@@ -400,21 +401,133 @@ pTransactionCalculateMinReqUTxO era' =
+          Opt.progDesc
+            "Connect to a running node to get context info and calculate the costs of the Plutus scripts of a given transaction."
+  )
+    <|> ( Opt.hsubparser $
```
same suggestion as above

palas replied:
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/ca0ab1fffb1b78353424921a47c30bdee50f9b56

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Transaction/Command.hs:
```diff
@@ -260,12 +265,50 @@ data TransactionCalculateMinValueCmdArgs era = TransactionCalculateMinValueCmdAr
+-- | Transaction context, required to evaluate the execution
+-- costs of the plutus scripts in the transaction.
+data TransactionContext era = TransactionContext
+  { shelleyBasedEra :: ShelleyBasedEra era
```
This field appears unused. Can we remove it? I think `TransactionContext` does not need to be parameterised on era.

`era` in `File (UTxO era) In` is just informative. If it gives us compilation issues, maybe `File (Some UTxO) In` would be a better fit here?

This could also make `TransactionCalculatePlutusScriptCostCmdArgs` era independent, so maybe it should belong to the `EraIndependent` commands hierarchy?

I see that this command depends only on the node's era and the transaction's era, so I don't see any reason why it shouldn't.

palas replied:
Where should we put it in the `EraIndependent` hierarchy? In `debug`?
The rest is addressed in: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/e25b52fcf1b340153f0618c49247d21e4287fbf0

carbolymer commented on cardano-cli/test/cardano-cli-test/Test/Cli/Transaction/Build.hs:
```diff
@@ -93,3 +126,87 @@ hprop_conway_transaction_build_raw_negative_bits_positive_total_txout = watchdog
+      , costOutFile
+      ]
+
+  json :: Aeson.Value <- H.evalEitherM $ H.readJsonFile costOutFile
```
```suggestion
  json :: Aeson.Value <- H.readJsonFileOk costOutFile
```

palas replied:
Applied here: https://github.com/IntersectMBO/cardano-cli/pull/1079/commits/5e415e4a4db68f009da20628971d2f55e513bafc
